homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,331 @@
+    return await component.async_unload_entry(entry)
+
+
+async def async_install(entity: UpdateEntity, service_call: ServiceCall) -> None:
+    """Service call wrapper to validate the call."""
+
+    # If version is not specified, but no update is available.
+    if (version := service_call.data.get(ATTR_VERSION)) is None and (
+        entity.current_version == entity.latest_version or entity.latest_version is None
+    ):
+        raise HomeAssistantError(f"No update available for {entity.name}")
+
+    # If version is specified, but not supported by the entity.
+    if version is not None and not entity.supported_features & SUPPORT_SPECIFIC_VERSION:
+        raise HomeAssistantError(
+            f"Installing a specific version is not supported for {entity.name}"
+        )
+
+    # If backup is requested, but not supported by the entity.
+    if (
+        backup := service_call.data.get(ATTR_BACKUP)
+    ) and not entity.supported_features & SUPPORT_BACKUP:
+        raise HomeAssistantError(f"Backup is not supported for {entity.name}")
+
+    # Update is already in progress.
+    if entity.in_progress is not False:
+        raise HomeAssistantError(
+            f"Update installation already in progress for {entity.name}"
+        )
+
+    await entity.async_install_with_progress(version, backup)
+
+
+@dataclass
+class UpdateEntityDescription(EntityDescription):
+    """A class that describes update entities."""
+
+    device_class: UpdateDeviceClass | str | None = None
+
+
+class UpdateEntity(RestoreEntity):
+    """Representation of an update entity."""
+
+    entity_description: UpdateEntityDescription
+    _attr_current_version: str | None = None
+    _attr_device_class: UpdateDeviceClass | str | None
+    _attr_in_progress: bool | int = False
+    _attr_latest_version: str | None = None
+    _attr_release_summary: str | None = None
+    _attr_release_url: str | None = None
+    _attr_state: None = None
+    _attr_supported_features: int = 0
+    _attr_title: str | None = None
+    __skipped_version: str | None = None
+    __in_progress: bool = False
+
+    @property
+    def current_version(self) -> str | None:
+        """Version currently in use."""
+        return self._attr_current_version
+
+    @property
+    def device_class(self) -> UpdateDeviceClass | str | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    def in_progress(self) -> bool | int | None:
+        """Update installation progress.
+
+        Needs SUPPORT_PROGRESS flag to be set for it to be used.
+
+        Can either return a boolean (True if in progress, False if not)
+        or an integer to indicate the progress in from 0 to 100%.
+        """
+        return self._attr_in_progress
```

Comment:
I wonder if this should return just a string instead? (Although strings quickly mean translations, while the current set doesn't need such a thing).